src/Native/System.Native/pal_networking.cpp

```diff
@@ -428,7 +429,168 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR
+#if !HAVE_GETHOSTBYNAME_R
+static int copy_hostent(struct hostent* from, struct hostent* to,
+                        char* buffer, size_t buflen)
+{
+    int i, len;
+    char* ptr = buffer;
+
+    to->h_addrtype = from->h_addrtype;
+    to->h_length = from->h_length;
+    to->h_name = static_cast<char*>(ptr);
+
+    /* copy hostname to buffer */
+    len = static_cast<int>(strlen(from->h_name)+1);
+    SafeStringCopy(ptr, buflen, from->h_name);
+    ptr += len;
+
+    if (strlen(ptr)+1 > buflen)
+    {
+        return ERANGE;
+    }
+
+    /* copy aliases to buffer */
+    to->h_aliases = reinterpret_cast<char**>(ptr);
+    for(i = 0; from->h_aliases[i]; i++);
+    ptr += (i+1) * static_cast<int>(sizeof(char*));
+
+    if (strlen(ptr)+1 > buflen)
```

Comment:
Why do we need to call strlen and walk the string again?  We know exactly how far ptr is from buffer.  Plus, strlen isn't even guaranteed to give the right length in this case. (I'm also not clear why we even moved ptr in this case... why did ptr advance after counting the number of aliases?